editoast/src/views/timetable/path_not_found_handler.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+
+use chrono::DateTime;
+use chrono::Utc;
+
+use crate::core::conflict_detection::ConflictDetectionRequest;
+use crate::core::conflict_detection::WorkSchedulesRequest;
+use crate::core::simulation::SimulationResponse;
+use crate::core::AsCoreRequest;
+use crate::core::CoreClient;
+use crate::error::Result;
+use crate::models::train_schedule::TrainSchedule;
+use crate::models::work_schedules::WorkSchedule;
+use crate::views::path::pathfinding::PathfindingResult;
+use crate::views::timetable::stdcm::STDCMResponse;
+
+use super::stdcm::build_train_requirements;
+
+pub struct PathNotFoundHandler {
+    pub core_client: Arc<CoreClient>,
+    pub infra_id: i64,
+    pub infra_version: String,
+    pub train_schedules: Vec<TrainSchedule>,
+    pub simulations: Vec<SimulationResponse>,
+    pub work_schedules: Vec<WorkSchedule>,
+    pub virtual_train_schedule: TrainSchedule,
+    pub virtual_train_sim_result: SimulationResponse,
+    pub virtual_train_pathfinding_result: PathfindingResult,
+    pub earliest_departure_time: DateTime<Utc>,
+    pub latest_simulation_end: DateTime<Utc>,
+}
+
+impl PathNotFoundHandler {
+    // Filter the provided work schedules to find those that conflict with the given parameters
+    // This identifies any work schedules that may overlap with the earliest departure time and latest simulation end.
+    fn get_work_schedules_request(&self) -> Option<WorkSchedulesRequest> {
+        if self.work_schedules.is_empty() {
+            return None;
+        }
+        let work_schedule_requirements = self
+            .work_schedules
+            .iter()
+            .filter_map(|ws| {
+                ws.get_id_and_core_work_schedule(
+                    self.earliest_departure_time,
+                    self.latest_simulation_end,
+                )
+            })
+            .collect();
+
+        Some(WorkSchedulesRequest {
+            start_time: self.earliest_departure_time,
+            work_schedule_requirements,
+        })
+    }
+
+    pub async fn handle(self) -> Result<STDCMResponse> {
+        let virtual_train_id = self.virtual_train_schedule.id;
+
+        // Combine the original train schedules with the virtual train schedule.
+        let train_schedules = [
+            self.train_schedules.clone(),
+            vec![self.virtual_train_schedule.clone()],
+        ]
+        .concat();
+
+        // Combine the original simulations with the virtual train's simulation results.
+        let simulations = [
+            self.simulations.clone(),
+            vec![self.virtual_train_sim_result.clone()],
+        ]
+        .concat();
```

Comment:
The function takes ownership of `self` and never returns it. I'd say it's fine to make it `mut` since it's already clear that the `SimulationFailureHandler` won't live after the call. Wdyt?

Reply:
I tested using `mut`, but it caused excessive cloning, so I tried destructuring instead. Let me know what you think.